From acfcdaf61b2322088fa94d7b66a00c8b1d67dbc1 Mon Sep 17 00:00:00 2001
From: Zhong Jinghua <zhongjinghua@huawei.com>
Date: Fri, 18 Nov 2022 17:06:58 +0800
Subject: [PATCH 0080/1398] [Huawei] scsi: fix scsi io performance degradation

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187946

----------------------------------------

We find commit 0c6a86b006d4 ("locking/atomic: Make test_and_*_bit()
ordered on failure") will degrade the fio performance of some drivers,
such as hisi_sas_v3.

Fixed it by using test_and_set_bit_relaxed(). This function is the
same as the test_and_set_bit() before the patch is incorporated.

Fixes: 0c6a86b006d4 ("locking/atomic: Make test_and_*_bit() ordered on failure")
Signed-off-by: Zhong Jinghua <zhongjinghua@huawei.com>
---
 drivers/scsi/scsi_lib.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/drivers/scsi/scsi_lib.c b/drivers/scsi/scsi_lib.c
index 9352a309fe55..92fe9151d492 100644
--- a/drivers/scsi/scsi_lib.c
+++ b/drivers/scsi/scsi_lib.c
@@ -52,6 +52,10 @@
 #define  SCSI_INLINE_SG_CNT  2
 #endif
 
+#ifndef test_and_set_bit_relaxed
+#define test_and_set_bit_relaxed test_and_set_bit
+#endif
+
 static struct kmem_cache *scsi_sense_cache;
 static struct kmem_cache *scsi_sense_isadma_cache;
 static DEFINE_MUTEX(scsi_sense_cache_mutex);
@@ -1597,7 +1601,7 @@ static void scsi_mq_done(struct scsi_cmnd *cmd)
 {
 	if (unlikely(blk_should_fake_timeout(cmd->request->q)))
 		return;
-	if (unlikely(test_and_set_bit(SCMD_STATE_COMPLETE, &cmd->state)))
+	if (unlikely(test_and_set_bit_relaxed(SCMD_STATE_COMPLETE, &cmd->state)))
 		return;
 	trace_scsi_dispatch_cmd_done(cmd);
 	blk_mq_complete_request(cmd->request);
-- 
2.17.1

